### ISSUE TITLE
Permalink: support via parameters

### ISSUE THREAD

**Issue by manuroe** [vector-im/riotX-android#287]
From riot-ios created by [manuroe](https://github.com/manuroe): vector-im/riot-ios#2547  

The spec is [MSC1704](https://github.com/matrix-org/matrix-doc/blob/master/proposals/1704-matrix.to-permalinks.md).

We need to:
- [ ] add them to permalinks the app creates
- [ ] pass them when joining a room
- [ ] pass them when joining an upgraded room

**Comment by ydnax** [comment]
are permalinks still a thing? currently room links on matrix.to  are unjoinable if the server in the rooms main address is down.
[This link](https://matrix.to/#/#fediventure:hackerspace.pl?via=hackerspace.pl&via=matrix.org&via=tchncs.de) for example does't work on either element android or desktop as long as hackerspace.pl is down and the users homeserver doesn't know the room yet.

**Comment by bmarty** [comment]
@ydnax if the homeserver is down and the users homeserver doesn't know the room yet, I think it is normal that you cannot join the room, hence I'm not a backend specialist.

**Comment by ydnax** [comment]
@bmarty from what i understand, the via parameters in the matrix.to are there for exactly this scenario: finding other servers if the main server is down, at least thats how i understand [MSC1704.](https://github.com/matrix-org/matrix-doc/blob/master/proposals/1704-matrix.to-permalinks.md)

**Comment by bmarty** [comment]
Yes, maybe. @turt2live WDYT?

**Comment by turt2live** [comment]
@bmarty this is recorded at https://matrix.org/docs/spec/appendices#routing

**Comment by bmarty** [comment]
This does not clarify what should happen if the server is down.

**Comment by turt2live** [comment]
Indeed. The client is only responsible for adding `via` parameters and converting them to `server_name` on join - the client is not responsible for ensuring the server is online.

**Comment by ydnax** [comment]
so this is a synapse server implementation issue and i shoud open an issue there as well?

**Comment by bmarty** [comment]
I do not know. Maybe.

Reading again this thread make me feel that I was not really accurate. Let's try to clarify.

### Scenario

- The client C is connected to a homeserver H1
- H1 is up
- The homeserver H2 hosts a room R (by hosts I mean: the room has been created on this server, the room id ends with the host name of H2)
- H2 is down
- C tries to join the room R

### Predicate

- C can only communicate with H1, and in no case directly to H2 or with any other homeserver.

So if H1 is down, the client cannot do anything.

- H1 can communicate with H2 or with any homeserver, if federation is enabled on H1 and on H2.

### So

If H2 is down, H1 cannot communicate with H2 and I assume could not let C join the room R.

I'm not sure but maybe if H1 already knows about R, it's possible to let C join the room even if H2 is down.

Once again, I'm not a specialist on the Matrix homeserver behavior. 

IIUC the via parameters are hosts for other homeservers knowing about R, since they are built using homeserver of joined members of the room R, and are used by H1 to communicate with other servers, if H1 does not already know about R, and if H2 is down.

### @ydnax link

https://matrix.to/#/#fediventure:hackerspace.pl?via=hackerspace.pl&via=matrix.org&via=tchncs.de

First problem: the first via parameter (`hackerspace.pl`) is useless because it's already in the room Id

If `matrix.org` and `tchncs.de` does not know about the room `#fediventure:hackerspace.pl`, and `hackerspace.pl` is down,  it's normal the the client cannot join the room.

If this link has been built by hand, it would explain why it's not working. There are maybe no member from `matrix.org` or `tchncs.de` in this room so these homeserver does not know about it.

**Comment by ydnax** [comment]
regarding the link: at the time i tried to join, it had members from both matrix.org or tchncs.de in it. the homeserver fror the account i tried to join had no users innthe room(i am quite sure as i am the only user on my homeserver). Another usuer from a fourth homeserver was able to join the room bit iirc someone else from the same homeserever was already in the room. this last part is only from memory and not 100% guaranteed as it was quite some time ago.
